### DISCUSSION TITLE
Authelia freezes the whole system at login - any way to debug?

### DISCUSSION THREAD

**Discussion by odwulf** [authelia/authelia#7002] [Troubleshooting] [answered]
Hello,

I have Authelia running through Docker as an authentication for LS.io Swag. I actually have three different systems using mostly the same config: a NAS, a Pi3, and a dedicated server. Configuration is kept simple as I am the only one to use those systems, so file backend it is.

On the NAS, it's smooth sailing. But on the Pi, very often (as in, it always take several tries to login), clicking the "Sign in" button just freezes everything, and I mean everything: other containers on the installation stop reaching health hooks, all logs stop, and I cannot log by SSH on the system. If I already am on the system,  everything is slow as to be unusable, though load average does no go further than 25 and ram is not entirely used. I notice the Authelia process using 25% CPU (so, the whole of one core?)

As I said, I am watching the logs, set to debug, but cannot see anything: it's like "Oh, we have someone trying to log on", then... two or three minutes(!) of absolutely nothing in the log while the whole system is stuck, before either a timeout (3 minutes for a request, that's a time out indeed ) and try your chance again, or something looking like this:

```
Unsuccessful 1FA authentication attempt by user 'USERNAME'" method=POST path=/api/firstfactor remote_ip=XXX.XXX.XXX.XXX stack="
	github.com/authelia/authelia/v4/internal/handlers/response.go:279
	markAuthenticationAttempt\ngithub.com/authelia/authelia/v4/internal/handlers/handler_firstfactor.go:60
	handleRouter.FirstFactorPOST.func10\ngithub.com/authelia/authelia/v4/internal/middlewares/bridge.go:
	handleRouter.(*BridgeBuilder).Build.func5.1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:65
	SecurityHeadersCSPNone.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:91
	SecurityHeadersNoStore.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:30
	SecurityHeadersBase.func1\ngithub.com/fasthttp/router@v1.5.0/router.go:441
	(*Router).Handler\ngithub.com/authelia/authelia/v4/internal/middlewares/log_request.go:14
	handleRouter.LogRequest.func40\ngithub.com/authelia/authelia/v4/internal/middlewares/strip_path.go:22
	handleRouter.StripPath.func41.1\ngithub.com/authelia/authelia/v4/internal/middlewares/errors.go:38
	RecoverPanic.func1\ngithub.com/valyala/fasthttp@v1.52.0/server.go:2374
	(*Server).serveConn\ngithub.com/valyala/fasthttp@v1.52.0/workerpool.go:224
	(*workerPool).workerFunc\ngithub.com/valyala/fasthttp@v1.52.0/workerpool.go:196
	(*workerPool).getCh.func1\nruntime/asm_arm64.s:1222
	goexit
"
```

And as I don't speak Go, I have no idea of what it can mean.

Are there other steps I can go through to try and find what's happening? Right now, I been making tentatives for 45 minutes, trying to restart Authelia and all, but I cannot pass.

(The same happens on the dedicated server, though not near as often, and is just an annoyance rather than an hindrance.)

**Comment by glime-ai** [comment]
Please note this response was created using generative AI leveraging previously available information from the repository. This information may be helpful however it is our strong recommendation that you carefully consider the information before blindly applying it.

---

It sounds like you're encountering a significant performance issue with Authelia on your Pi3, which is affecting not only the authentication process but also the overall system stability. The fact that the system becomes nearly unusable and other containers are affected suggests that the issue might be related to the resources available on the Pi3, especially considering that Authelia is observed to be using 25% CPU during these incidents. This could indicate that Authelia is hitting a resource bottleneck, particularly on a less powerful device like the Pi3 compared to your NAS or dedicated server.

Given the symptoms you've described, here are a few steps and considerations to help diagnose and potentially resolve the issue:

1. **Resource Constraints**: The Raspberry Pi 3 has limited resources compared to a NAS or a dedicated server. Authelia's performance, especially when using file-based backends and running alongside other containers, might be constrained by the available CPU and memory. Monitoring system resources closely during login attempts could provide insights into potential bottlenecks.

2. **Logs and Debugging**: Since you've already set the logs to debug without much insight, consider checking not just Authelia's logs but also system logs (`/var/log/syslog`, `/var/log/messages`) for any anomalies or errors related to system resources or Docker itself during the time of the issue.

3. **Configuration Review**: Double-check your Authelia configuration files across the different systems to ensure there are no unintended differences, especially those that might increase resource usage on the Pi3. Pay particular attention to session and storage providers, as using more resource-intensive options could exacerbate the issue on less powerful hardware.

4. **Software Versions**: Ensure all related software (Docker, Authelia, and any dependencies) are up to date on the Pi3. Sometimes, performance issues or bugs are resolved in newer versions.

5. **Performance Tuning**: Consider tuning Authelia's performance settings if available. For example, adjusting session expiration times, garbage collection intervals, or even the hashing algorithm's difficulty (if using local file authentication) could potentially reduce resource usage.

6. **Alternative Hardware**: While not always an option, testing Authelia on a Raspberry Pi 4 or another more powerful device could help determine if the issue is strictly related to the hardware limitations of the Pi3.

7. **Community and Support**: If the issue persists, consider reaching out to the Authelia community or support channels with detailed information about your setup, configuration, and the symptoms you're experiencing. The Authelia GitHub discussions or issues sections could be a good place to start.

Given the complexity of the issue and the potential for it to be related to a combination of factors including hardware limitations, software configuration, and possibly bugs, a process of elimination and careful monitoring will be key to diagnosing and resolving the problem.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=6fsHkcyufjQLaB&utm_medium=github).

**Comment by james-d-elliott** [comment]
Can you show your authentication backend settings? Also how much RAM do you have available without it running?

**Comment by odwulf** [comment]
Here is is:
```
authentication_backend:
  file:
    path: /config/users_database.yml
    password:
      # algorithm: argon2id
      algorithm: sha512
      iterations: 1000
      key_length: 32
      salt_length: 16
      memory: 32
      parallelism: 8
```

Also, it's a 1Gb RAM system. Which, yeah, can be a bit on the low end. But as I don't have that much problem on a system which only has two and is running Clamav, I still had hopes.

**Reply by odwulf** [reply]
(And I'll see if I can try authenticating on a (separate) LDAP server, to check if it really is an authentication problem.)

**Reply by james-d-elliott** [reply]
Very unusual, that should be fine. What do the password hashes look like? Just the information starting with the first `$` and ending with the second `$`.

**Reply by odwulf** [reply]
Holy cow: `$argon2id$` So it really is an error on my part. I switched the config to sha512 months ago because of this problem, and never thought about rehashing the password. Does it mean that it was still using Argon2 to check it?

In the meantime, I set up a LDAP server to try it as a backend for Authelia and... logging in is instant.

**Reply by james-d-elliott** [reply, marked as answer]
Yeah it does, only new hashes (i.e. reset password) will be generated using new settings. Since we don't know the raw input we can't regenerate them and it'd be poor practice to update it on login at least in my opinion.

**Reply by nightah** [reply]
Yes, the passwords need to be rehashed when you change algorithms.
The hashes with `$argon2id$` would have still been using whatever values you used when that was initially hashed (you should be able to tell by what comes after the hash header).

**Reply by odwulf** [reply]
Obviously you are right, it was a huge oversight on my part.

I just retried with a newly hashed password, and though it is a tad slower (like 5secs rather than 2) than with LDAP, it was clearly where the problem was coming for.

I did not make tests with different algorithms back then, as changing the config did not solve my problem (DUH) and it is not a sensitive system anyway, but now that I have a better understanding, I definitely will.

Thanks for your very quick and efficient help, and sorry for the noise. Keep up the good work. Case closed.

**Reply by james-d-elliott** [reply]
A lot of LDAP backends use single-iteration hashing algorithms. You can tune ours to your liking, we suggest roughly 1 second is good for most environments.

**Reply by odwulf** [reply]
There's definitely room for wiggling here, as I was able to run the LDAP server as a package on a more powerful machine. The Docker/Swag/Authelia stack on a PI is little more than a entry door in my case (but the only appliance on that site I have complete control over).
